test/e2e/operator_test.go

```diff
@@ -228,3 +233,402 @@ func setImage(deployment *appsv1.Deployment, image string) {
+		if ver.Name == "coredns" {
+			if len(ver.Version) == 0 {
+				t.Fatalf("dns operator does not contain a coredns version: %v", err)
+			}
+			upstreamImage = ver.Version
+			break
+		}
+	}
+
+	// Create the upstream resolver Pod.
+	upstreamResolver := upstreamPod(upstreamPodName, upstreamPodNS, upstreamImage, upstreamPodName)
+	if err := cl.Create(context.TODO(), upstreamResolver); err != nil {
+		t.Fatalf("failed to create pod %s/%s: %v", upstreamResolver.Namespace, upstreamResolver.Name, err)
+	}
+	defer func() {
+		if err := cl.Delete(context.TODO(), upstreamResolver); err != nil {
+			t.Fatalf("failed to delete pod %s/%s: %v", upstreamResolver.Namespace, upstreamResolver.Name, err)
+		}
+	}()
+
+	// Wait for the upstream resolver Pod to be ready.
+	err = wait.PollImmediate(1*time.Second, 60*time.Second, func() (bool, error) {
+		if err := cl.Get(context.TODO(), types.NamespacedName{Namespace: upstreamResolver.Namespace, Name: upstreamResolver.Name}, upstreamResolver); err != nil {
+			return false, nil
+		}
+		for _, cond := range upstreamResolver.Status.Conditions {
+			if cond.Type == corev1.ContainersReady &&
+				cond.Status == corev1.ConditionTrue {
+				return true, nil
+			}
+		}
+		return false, nil
+	})
+	if err != nil {
+		t.Fatalf("failed to observe ContainersReady condition for pod %s/%s: %v", upstreamResolver.Namespace, upstreamResolver.Name, err)
+	}
+
+	// Create the upstream resolver Service and get the ClusterIP.
+	upstreamSvc := upstreamService(upstreamPodName, upstreamPodNS)
+	if err := cl.Create(context.TODO(), upstreamSvc); err != nil {
+		t.Fatalf("failed to create service %s/%s: %v", upstreamSvc.Namespace, upstreamSvc.Name, err)
+	}
+	defer func() {
+		if err := cl.Delete(context.TODO(), upstreamSvc); err != nil {
+			t.Fatalf("failed to delete service %s/%s: %v", upstreamSvc.Namespace, upstreamSvc.Name, err)
+		}
+	}()
+	if err := cl.Get(context.TODO(), types.NamespacedName{upstreamSvc.Namespace, upstreamSvc.Name}, upstreamSvc); err != nil {
+		t.Fatalf("failed to get service %s/%s: %v", upstreamSvc.Namespace, upstreamSvc.Name, err)
+	}
+	upstreamIP := upstreamSvc.Spec.ClusterIP
+
+	// Update cluster DNS forwarding with the upstream resolver's Service IP address.
+	defaultDNS := &operatorv1.DNS{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: "openshift-dns",
+			Name:      "default",
+		},
+	}
+	if err := cl.Get(context.TODO(), types.NamespacedName{defaultDNS.Namespace, defaultDNS.Name}, defaultDNS); err != nil {
+		t.Fatalf("failed to get dns %s/%s: %v", defaultDNS.Namespace, defaultDNS.Name, err)
+	}
+	upstream := operatorv1.Server{
+		Name:  "test",
+		Zones: []string{"foo.com"},
+		ForwardPlugin: operatorv1.ForwardPlugin{
+			Upstreams: []string{upstreamIP},
+		},
+	}
+	defaultDNS.Spec.Servers = []operatorv1.Server{upstream}
+	if err := cl.Update(context.TODO(), defaultDNS); err != nil {
+		t.Fatalf("failed to update dns %s/%s: %v", defaultDNS.Namespace, defaultDNS.Name, err)
+	}
```

Comment:
Need to ensure that the default dns resource is always reverted to its original state.

Reply:
ack